### ISSUE TITLE
RxAbsListView scroll listener binding

### ISSUE THREAD

**Issue by JakeWharton** [JakeWharton/RxBinding#80]

**Comment by skoric** [comment]
Is there a good way to create an event that can actually hold two different sets of variables, depending on the invoked listener method? For example, the case with `SnackbarDismissesOnSubscribe`. How could we create `SnackbarShowsOnSubscribe` since they would share the listener?

Would something like this be considered a bad design, by getting the emitted (sub)event through a null check?
```
package com.jakewharton.rxbinding.widget;

import android.content.Context;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.widget.AbsListView;
import com.jakewharton.rxbinding.view.ViewEvent;

/**
 * A scroll event on {@link AbsListView}.
 * <p>
 * <strong>Warning:</strong> Instances keep a strong reference to the view. Operators that cache
 * instances have the potential to leak the associated {@link Context}.
 */
public final class AbsListViewScrollEvent extends ViewEvent<AbsListView> {
  @CheckResult @NonNull
  public static AbsListViewScrollEvent create(@NonNull AbsListView view, int firstVisibleItem,
      int visibleItemCount, int totalItemCount) {
    return new AbsListViewScrollEvent(view, firstVisibleItem, visibleItemCount, totalItemCount);
  }

  @CheckResult @NonNull
  public static AbsListViewScrollEvent create(@NonNull AbsListView view, int scrollState) {
    return new AbsListViewScrollEvent(view, scrollState);
  }

  @Nullable private final OnScroll onScroll;
  @Nullable private final OnScrollStateChanged onScrollStateChanged;

  private AbsListViewScrollEvent(@NonNull AbsListView view, int firstVisibleItem,
      int visibleItemCount, int totalItemCount) {
    super(view);
    this.onScroll = OnScroll.create(view, firstVisibleItem, visibleItemCount, totalItemCount);
    this.onScrollStateChanged = null;
  }

  private AbsListViewScrollEvent(@NonNull AbsListView view, int scrollState) {
    super(view);
    this.onScroll = null;
    this.onScrollStateChanged = OnScrollStateChanged.create(view, scrollState);
  }

  @Nullable public OnScroll onScroll() {
    return onScroll;
  }

  @Nullable public OnScrollStateChanged onScrollStateChanged() {
    return onScrollStateChanged;
  }

  @Override public boolean equals(Object o) {
    if (o == this) return true;
    if (!(o instanceof AbsListViewScrollEvent)) return false;
    AbsListViewScrollEvent other = (AbsListViewScrollEvent) o;
    return other.view() == view()
        && onScroll == other.onScroll
        && onScrollStateChanged == other.onScrollStateChanged;
  }

  @Override public int hashCode() {
    int result = 17;
    result = result * 37 + view().hashCode();
    result = result * 37 + onScroll.hashCode();
    result = result * 37 + onScrollStateChanged.hashCode();
    return result;
  }

  @Override public String toString() {
    return "AbsListViewScrollEvent{" +
        "onScroll=" + onScroll +
        ", onScrollStateChanged=" + onScrollStateChanged +
        '}';
  }

  private static final class OnScroll extends ViewEvent<AbsListView> {
    @CheckResult @NonNull
    private static OnScroll create(@NonNull AbsListView view, int firstVisibleItem,
        int visibleItemCount, int totalItemCount) {
      return new OnScroll(view, firstVisibleItem, visibleItemCount, totalItemCount);
    }

    private final int firstVisibleItem;
    private final int visibleItemCount;
    private final int totalItemCount;

    private OnScroll(@NonNull AbsListView view, int firstVisibleItem,
        int visibleItemCount, int totalItemCount) {
      super(view);
      this.firstVisibleItem = firstVisibleItem;
      this.visibleItemCount = visibleItemCount;
      this.totalItemCount = totalItemCount;
    }

    public int firstVisibleItem() {
      return firstVisibleItem;
    }

    public int visibleItemCount() {
      return visibleItemCount;
    }

    public int totalItemCount() {
      return totalItemCount;
    }

    @Override public boolean equals(Object o) {
      if (o == this) return true;
      if (!(o instanceof AbsListViewScrollEvent.OnScroll)) return false;
      AbsListViewScrollEvent.OnScroll other = (AbsListViewScrollEvent.OnScroll) o;
      return other.view() == view()
          && firstVisibleItem == other.firstVisibleItem
          && visibleItemCount == other.visibleItemCount
          && totalItemCount == other.totalItemCount;
    }

    @Override public int hashCode() {
      int result = 17;
      result = result * 37 + view().hashCode();
      result = result * 37 + firstVisibleItem;
      result = result * 37 + visibleItemCount;
      result = result * 37 + totalItemCount;
      return result;
    }

    @Override public String toString() {
      return "OnScroll{" +
          "firstVisibleItem=" + firstVisibleItem +
          ", visibleItemCount=" + visibleItemCount +
          ", totalItemCount=" + totalItemCount +
          '}';
    }
  }

  private static final class OnScrollStateChanged extends ViewEvent<AbsListView> {
    @CheckResult @NonNull
    private static OnScrollStateChanged create(@NonNull AbsListView view, int scrollState) {
      return new OnScrollStateChanged(view, scrollState);
    }

    private final int scrollState;

    private OnScrollStateChanged(@NonNull AbsListView view, int scrollState) {
      super(view);
      this.scrollState = scrollState;
    }

    public int scrollState() {
      return scrollState;
    }

    @Override public boolean equals(Object o) {
      if (o == this) return true;
      if (!(o instanceof AbsListViewScrollEvent.OnScrollStateChanged)) return false;
      AbsListViewScrollEvent.OnScrollStateChanged other =
          (AbsListViewScrollEvent.OnScrollStateChanged) o;
      return other.view() == view()
          && scrollState == other.scrollState;
    }

    @Override public int hashCode() {
      int result = 17;
      result = result * 37 + view().hashCode();
      result = result * 37 + scrollState;
      return result;
    }

    @Override public String toString() {
      return "OnScroll{" +
          "scrollState=" + scrollState +
          '}';
    }
  }
}
```

**Comment by JakeWharton** [comment]
Merged as a201d000598c59b7ba4268f4f7625a1fa9d331b3.
